diff --git a/src/esp/assets/ResourceManager.cpp b/src/esp/assets/ResourceManager.cpp
index cf74114..62e6a97 100644
--- a/src/esp/assets/ResourceManager.cpp
+++ b/src/esp/assets/ResourceManager.cpp
@@ -1614,7 +1614,7 @@ gfx::PhongMaterialData::uptr ResourceManager::buildPhongShadedMaterialData(
   }
 
   // specular material properties
-  finalMaterial->specularColor = material.specularColor();
+  finalMaterial->specularColor = 0x00000000_rgbaf;
   if (material.flags() & Mn::Trade::PhongMaterialData::Flag::SpecularTexture) {
     finalMaterial->specularTexture =
         textures_[textureBaseIndex + material.specularTexture()].get();
