// optimal/esm/StringBuilder.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";

/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */
import Builder from './Builder';

function isString(value) {
  return typeof value === 'string' && value !== '';
}

var StringBuilder = function (_Builder) {
  _inherits(StringBuilder, _Builder);

  function StringBuilder() {
    var _this;

    var defaultValue = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';

    _classCallCheck(this, StringBuilder);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(StringBuilder).call(this, 'string', defaultValue));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "allowEmpty", false);

    if ("production" !== process.env.NODE_ENV) {
      _this.addCheck(_this.checkNotEmpty);
    }

    return _this;
  }

  _createClass(StringBuilder, [{
    key: "contains",
    value: function contains(token) {
      var index = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;

      if ("production" !== process.env.NODE_ENV) {
        this.invariant(isString(token), 'Contains requires a non-empty string.');
      }

      return this.addCheck(this.checkContains, token, index);
    }
  }, {
    key: "checkContains",
    value: function checkContains(path, value, token) {
      var index = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 0;

      if ("production" !== process.env.NODE_ENV) {
        this.invariant(value.indexOf(token, index) >= 0, "String does not include \"".concat(token, "\"."), path);
      }
    }
  }, {
    key: "match",
    value: function match(pattern) {
      if ("production" !== process.env.NODE_ENV) {
        this.invariant(pattern instanceof RegExp, 'Match requires a regular expression to match against.');
      }

      return this.addCheck(this.checkMatch, pattern);
    }
  }, {
    key: "checkMatch",
    value: function checkMatch(path, value, pattern) {
      if ("production" !== process.env.NODE_ENV) {
        this.invariant(value.match(pattern), "String does not match pattern \"".concat(pattern.source, "\"."), path);
      }
    }
  }, {
    key: "empty",
    value: function empty() {
      if ("production" !== process.env.NODE_ENV) {
        this.allowEmpty = true;
      }

      return this;
    }
  }, {
    key: "checkNotEmpty",
    value: function checkNotEmpty(path, value) {
      if ("production" !== process.env.NODE_ENV) {
        if (!this.allowEmpty) {
          this.invariant(isString(value), 'String cannot be empty.', path);
        }
      }
    }
  }, {
    key: "oneOf",
    value: function oneOf(list) {
      if ("production" !== process.env.NODE_ENV) {
        this.invariant(Array.isArray(list) && list.length > 0 && list.every(function (item) {
          return isString(item);
        }), 'One of requires a non-empty array of strings.');
      }

      return this.addCheck(this.checkOneOf, list);
    }
  }, {
    key: "checkOneOf",
    value: function checkOneOf(path, value, list) {
      if ("production" !== process.env.NODE_ENV) {
        this.invariant(list.indexOf(value) >= 0, "String must be one of: ".concat(list.join(', ')), path);
      }
    }
  }]);

  return StringBuilder;
}(Builder);

export { StringBuilder as default };
export function string() {
  var defaultValue = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
  return new StringBuilder(defaultValue);
}
