// optimal/lib/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Builder: true,
  bool: true,
  custom: true,
  func: true,
  CollectionBuilder: true,
  array: true,
  object: true,
  InstanceBuilder: true,
  instance: true,
  date: true,
  regex: true,
  NumberBuilder: true,
  number: true,
  ShapeBuilder: true,
  shape: true,
  StringBuilder: true,
  string: true,
  UnionBuilder: true,
  union: true
};
Object.defineProperty(exports, "Builder", {
  enumerable: true,
  get: function get() {
    return _Builder.default;
  }
});
Object.defineProperty(exports, "bool", {
  enumerable: true,
  get: function get() {
    return _Builder.bool;
  }
});
Object.defineProperty(exports, "custom", {
  enumerable: true,
  get: function get() {
    return _Builder.custom;
  }
});
Object.defineProperty(exports, "func", {
  enumerable: true,
  get: function get() {
    return _Builder.func;
  }
});
Object.defineProperty(exports, "CollectionBuilder", {
  enumerable: true,
  get: function get() {
    return _CollectionBuilder.default;
  }
});
Object.defineProperty(exports, "array", {
  enumerable: true,
  get: function get() {
    return _CollectionBuilder.array;
  }
});
Object.defineProperty(exports, "object", {
  enumerable: true,
  get: function get() {
    return _CollectionBuilder.object;
  }
});
Object.defineProperty(exports, "InstanceBuilder", {
  enumerable: true,
  get: function get() {
    return _InstanceBuilder.default;
  }
});
Object.defineProperty(exports, "instance", {
  enumerable: true,
  get: function get() {
    return _InstanceBuilder.instance;
  }
});
Object.defineProperty(exports, "date", {
  enumerable: true,
  get: function get() {
    return _InstanceBuilder.date;
  }
});
Object.defineProperty(exports, "regex", {
  enumerable: true,
  get: function get() {
    return _InstanceBuilder.regex;
  }
});
Object.defineProperty(exports, "NumberBuilder", {
  enumerable: true,
  get: function get() {
    return _NumberBuilder.default;
  }
});
Object.defineProperty(exports, "number", {
  enumerable: true,
  get: function get() {
    return _NumberBuilder.number;
  }
});
Object.defineProperty(exports, "ShapeBuilder", {
  enumerable: true,
  get: function get() {
    return _ShapeBuilder.default;
  }
});
Object.defineProperty(exports, "shape", {
  enumerable: true,
  get: function get() {
    return _ShapeBuilder.shape;
  }
});
Object.defineProperty(exports, "StringBuilder", {
  enumerable: true,
  get: function get() {
    return _StringBuilder.default;
  }
});
Object.defineProperty(exports, "string", {
  enumerable: true,
  get: function get() {
    return _StringBuilder.string;
  }
});
Object.defineProperty(exports, "UnionBuilder", {
  enumerable: true,
  get: function get() {
    return _UnionBuilder.default;
  }
});
Object.defineProperty(exports, "union", {
  enumerable: true,
  get: function get() {
    return _UnionBuilder.union;
  }
});
exports.default = void 0;

var _optimal = _interopRequireDefault(require("./optimal"));

var _Builder = _interopRequireWildcard(require("./Builder"));

var _CollectionBuilder = _interopRequireWildcard(require("./CollectionBuilder"));

var _InstanceBuilder = _interopRequireWildcard(require("./InstanceBuilder"));

var _NumberBuilder = _interopRequireWildcard(require("./NumberBuilder"));

var _ShapeBuilder = _interopRequireWildcard(require("./ShapeBuilder"));

var _StringBuilder = _interopRequireWildcard(require("./StringBuilder"));

var _UnionBuilder = _interopRequireWildcard(require("./UnionBuilder"));

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _types[key];
    }
  });
});

/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */
var _default = _optimal.default;
exports.default = _default;
