// optimal/lib/InstanceBuilder.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.instance = instance;
exports.regex = regex;
exports.date = date;
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _Builder2 = _interopRequireDefault(require("./Builder"));

var _isObject = _interopRequireDefault(require("./isObject"));

/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */
var InstanceBuilder = function (_Builder) {
  (0, _inherits2.default)(InstanceBuilder, _Builder);

  function InstanceBuilder() {
    var _this;

    var refClass = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
    (0, _classCallCheck2.default)(this, InstanceBuilder);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(InstanceBuilder).call(this, 'instance', null));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)((0, _assertThisInitialized2.default)(_this)), "refClass", null);

    _this.nullable();

    if ("production" !== process.env.NODE_ENV) {
      if (refClass) {
        _this.invariant(typeof refClass === 'function', 'A class reference is required.');
      }

      _this.refClass = refClass;

      _this.addCheck(_this.checkInstance, refClass);
    }

    return _this;
  }

  (0, _createClass2.default)(InstanceBuilder, [{
    key: "checkInstance",
    value: function checkInstance(path, value, refClass) {
      if ("production" !== process.env.NODE_ENV) {
        if (refClass) {
          this.invariant(typeof refClass === 'function' && value instanceof refClass, "Must be an instance of \"".concat(this.typeAlias(), "\"."), path);
        } else {
          this.invariant((0, _isObject.default)(value) && value.constructor !== Object, 'Must be a class instance.', path);
        }
      }
    }
  }, {
    key: "typeAlias",
    value: function typeAlias() {
      var refClass = this.refClass;
      return refClass ? refClass.name || refClass.constructor.name : 'class';
    }
  }]);
  return InstanceBuilder;
}(_Builder2.default);

exports.default = InstanceBuilder;

function instance() {
  var refClass = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
  return new InstanceBuilder(refClass);
}

function regex() {
  return instance(RegExp);
}

function date() {
  return instance(Date);
}
